### DISCUSSION TITLE
IDE0057: Slice can be simplified

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#3879] [General] [answered]
Some C# analyzer tells me that this line can be optimized:

`new Vector<T>(B.Slice(Idx))`

IDE0057: Slice can be simplified to this

`new Vector<T>(B[Idx..])`

What is the benefit only less text. I would think a range is created here that slows down performance?

**Comment by YairHalberstadt** [comment, marked as answer]
> What is the benefit only less text.

Correct, although I would argue it's more that's it's more declarative than less text. I don't know what the parameters to slice are. Length? Start? Finish? But ranges are a well known construct.

> I would think a range is created here that slows down performance?

The runtime team works very hard to make sure that isn't the case. They definitely will treat it as a bug if the Jit output is worse.

**Reply by msedi** [reply]
@YairHalberstadt. OK Thanks. If there's no performance drawback I would go with this. But it wouldn't help me if  I vectorize, but get a slowdown when using the Range ;-)

**Reply by svick** [reply]
@YairHalberstadt The output is a bit worse, at least based on [what I see on SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGvQYBlABbYoABwAy2YCwBKHAHYZe3YdRHkkTUgwDCNAN40GVhgG0AUrwwBxGGp78AFBgCeMmBAAzD14NAEpQgF1La2YDYhRxABt+GAAeABUAPg8xGWw1DMyGACE0BhCMBgBJABMEUIYAdwkeGAZ0hhAGXAwoDjBKi2prEYY3RoYANRgB6EKPYpYxZLAYD1r60J0RgF9oq33bB2dXdzAvX38givCo4Zj9JgTlfIBzNKycvIKskrKK6p1BrNVrtTrdXr9QaHEbjKYzDBzT7FGwbFgsCJbQ57ag7IA===).

**Reply by msedi** [reply]
@svick: Should I file it as an issue?

**Reply by YairHalberstadt** [reply]
@msedi yes

**Reply by YairHalberstadt** [reply]
@svick make sure you're checking .Net 5, not X64 which is .net core 3.1

**Reply by BreyerW** [reply]
Also non-span based range should be considerably worse than span. But thats unfixable, net 5 even introduced an analyzer to warn on non-span based range which is annoying but c'est la vie (normally non-span based range should behave like span-based but due to limits that isnt the case)

**Comment by CyrusNajmabadi** [comment]
There's no issue here with the analyzer fwiw.  The analyzer is just telling you can simplify things in this manner if you so choose.

If the jitted code is suboptimal, you can certainly open an issue on that.

**Reply by msedi** [reply]
See https://github.com/dotnet/roslyn/issues/47626
